apollo-federation/src/query_plan/optimize.rs

```diff
@@ -765,3 +825,420 @@ impl SelectionSet {
+    ) -> Result<SelectionSet, FederationError> {
+        if fragments.is_empty() {
+            return Ok(self.clone());
+        }
+
+        // Calling optimize() will not match a fragment that would have expanded at
+        // top-level. That is, say we have the selection set `{ x y }` for a top-level `Query`, and
+        // we have a fragment
+        // ```
+        // fragment F on Query {
+        //   x
+        //   y
+        // }
+        // ```
+        // then calling `self.optimize(fragments)` would only apply check if F apply to
+        // `x` and then `y`.
+        //
+        // To ensure the fragment match in this case, we "wrap" the selection into a trivial
+        // fragment of the selection parent, so in the example above, we create selection `... on
+        // Query { x y}`. With that, `optimize` will correctly match on the `on Query`
+        // fragment; after which we can unpack the final result.
+        let wrapped = InlineFragmentSelection::from_selection_set(
+            self.type_position.clone(), // parent type
+            self.clone(),               // selection set
+            Default::default(),         // directives
+        );
+        let mut validator = FieldsConflictMultiBranchValidator::from_initial_validator(
+            FieldsConflictValidator::from_selection_set(self),
+        );
+        let optimized = wrapped.optimize(fragments, &mut validator)?;
+
+        // Now, it's possible we matched a full fragment, in which case `optimized` will be just
+        // the named fragment, and in that case we return a singleton selection with just that.
+        // Otherwise, it's our wrapping inline fragment with the sub-selections optimized, and we
+        // just return that subselection.
+
+        match optimized {
+            Selection::FragmentSpread(fragment) => Ok(SelectionSet::make_selection_set(
+                &self.schema,
+                &self.type_position, // parent type
+                std::iter::once(fragment.selection_set.selections.values()),
+                fragments,
+            )?),
+
+            _ => optimized
+                .selection_set()?
+                .ok_or_else(|| {
+                    FederationError::internal(
+                        "Unexpected: missing sub-selection set after optimization",
+                    )
+                })
+                .cloned(),
+        }
+    }
+}
+
+impl Operation {
+    const DEFAULT_MIN_USAGES_TO_OPTIMIZE: u32 = 2;
+
+    pub(crate) fn optimize(
+        &mut self,
+        fragments: Option<&NamedFragments>,
+        min_usages_to_optimize: Option<u32>,
+    ) -> Result<(), FederationError> {
+        let min_usages_to_optimize =
+            min_usages_to_optimize.unwrap_or(Self::DEFAULT_MIN_USAGES_TO_OPTIMIZE);
+        let Some(fragments) = fragments else {
+            return Ok(());
+        };
+        if fragments.is_empty() {
+            return Ok(());
+        }
+        assert!(
```

Comment:
We can avoid this assert either by taking a `Option<NonZeroU32>` parameter, or maybe even removing it entirely (AFAIK JS only used it in tests)

Reply:
Makes sense. I'll remove the parameter altogether.